### ISSUE TITLE
false positive in MemorySanitizer when writes are done in a FFI call

### ISSUE THREAD

**Issue by BusyJay** [rust-lang/rust#75640]
I tried this code:

```rust
fn main() {
    unsafe {
        let src = b"";
        let bound = lz4_sys::LZ4_compressBound(src.len() as i32);
        let mut output = Vec::<u8>::with_capacity(bound as usize);
        // If uncomment following for loop, assert will pass.
        //for _ in 0..bound {
        //    output.push(0xff);
        //}
        let size = lz4_sys::LZ4_compress_default(
            src.as_ptr() as _,
            output.as_mut_ptr() as _,
            src.len() as i32,
            bound,
        );
        assert!(size <= bound);
        output.set_len(size as usize);
        for v in &output {
            assert_eq!(*v, 0);
        }
    }
}
```

I expected to see this happen:
Running with memory sanitizer should exit normally.

Instead, this happened: 
```
env RUSTFLAGS='-Zsanitizer=memory -Zsanitizer-memory-track-origins' cargo +nightly run -Zbuild-std --target x86_64-unknown-linux-gnu
==13035==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x55a4629d2ffa in test_false_positive::main::h135f8e367b9ee96a (/mnt/workspace/test-false-positive/target/x86_64-unknown-linux-gnu/debug/test-false-positive+0x77ffa)
    #1 0x55a4629d4125 in std::rt::lang_start::_$u7b$$u7b$closure$u7d$$u7d$::he5a302ed29938521 (/mnt/workspace/test-false-positive/target/x86_64-unknown-linux-gnu/debug/test-false-positive+0x79125)
    #2 0x55a462a03057 in std::rt::lang_start_internal::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::h0d9989ce28f66057 (/mnt/workspace/test-false-positive/target/x86_64-unknown-linux-gnu/debug/test-false-positive+0xa8057)
    #3 0x55a462ae5464 in std::sys_common::backtrace::__rust_begin_short_backtrace::h2cc322a62bb356db (/mnt/workspace/test-false-positive/target/x86_64-unknown-linux-gnu/debug/test-false-positive+0x18a464)
    #4 0x55a462a03455 in std::rt::lang_start_internal::_$u7b$$u7b$closure$u7d$$u7d$::h2cc1cb770cf40ea4 (/mnt/workspace/test-false-positive/target/x86_64-unknown-linux-gnu/debug/test-false-positive+0xa8455)
    #5 0x55a462a510f9 in std::panicking::try::do_call::hce2558f4571e39c8 (/mnt/workspace/test-false-positive/target/x86_64-unknown-linux-gnu/debug/test-false-positive+0xf60f9)
    #6 0x55a462a573ab in __rust_try (/mnt/workspace/test-false-positive/target/x86_64-unknown-linux-gnu/debug/test-false-positive+0xfc3ab)
    #7 0x55a462a4ff34 in std::panicking::try::h9b50d2f4190a0e96 (/mnt/workspace/test-false-positive/target/x86_64-unknown-linux-gnu/debug/test-false-positive+0xf4f34)
    #8 0x55a462a46f7e in std::panic::catch_unwind::h169c903f4c0ff425 (/mnt/workspace/test-false-positive/target/x86_64-unknown-linux-gnu/debug/test-false-positive+0xebf7e)
    #9 0x55a462a02b91 in std::rt::lang_start_internal::hf0abd1f790105145 (/mnt/workspace/test-false-positive/target/x86_64-unknown-linux-gnu/debug/test-false-positive+0xa7b91)
    #10 0x55a4629d3fc7 in std::rt::lang_start::h326dcedc286951a9 (/mnt/workspace/test-false-positive/target/x86_64-unknown-linux-gnu/debug/test-false-positive+0x78fc7)
    #11 0x55a4629d39bc in main (/mnt/workspace/test-false-positive/target/x86_64-unknown-linux-gnu/debug/test-false-positive+0x789bc)
    #12 0x7f4b46460554 in __libc_start_main (/lib64/libc.so.6+0x22554)
    #13 0x55a462974a98 in _start (/mnt/workspace/test-false-positive/target/x86_64-unknown-linux-gnu/debug/test-false-positive+0x19a98)
```
It failed when running the assert check in the last for loop.

However, if I uncommented the first for loop in above code, the program could exit normally. This seemed to be conflict with what the sanitizer suggested. If there was any variable uninitialized, the assert should fail as the for loop set all values to 0xff.

I suspected there were some special tricks done in `LZ4_compress_default` that misleaded memory sanitizer.

### Meta

`rustc --version --verbose`:
```
rustc 1.46.0-nightly (346aec9b0 2020-07-11)
binary: rustc
commit-hash: 346aec9b02f3c74f3fce97fd6bda24709d220e49
commit-date: 2020-07-11
host: x86_64-unknown-linux-gnu
release: 1.46.0-nightly
LLVM version: 10.0
```

**Comment by BusyJay** [comment]
Reproduced it with latest nightly:
```
rustc 1.47.0-nightly (7e6d6e5f5 2020-08-16)
binary: rustc
commit-hash: 7e6d6e5f535321c2223f044caba16f97b825009c
commit-date: 2020-08-16
host: x86_64-unknown-linux-gnu
release: 1.47.0-nightly
LLVM version: 10.0
```

**Comment by alex** [comment]
Did you compile lz4 with msan?

Unlike ASAN/UBSAN, for MSAN, _all_ code linked into the process needs to be compiled with MSAN. You can't use normal distro libraries for things.

**Comment by BusyJay** [comment]
I'm not sure. Is there anything needs to be taken care of by cc? It seems it just invokes cc normally.  https://github.com/10XGenomics/lz4-rs/blob/master/lz4-sys/build.rs

**Comment by alex** [comment]
AFAIK, `cc` does not automatically pass the sanitizer used to compile Rust code down to the compiler used to compile C code. Whether this is a bug in `cc` or in every package that uses `cc` is someone who isn't me's decision.

**Comment by tmiasko** [comment]
The cc crate includes CC & CFLAGS, so for vendored C dependencies following should work (as long as clang is based on compatible LLVM version):

```shell
export \
  'CC=clang'\
  'CFLAGS=-fsanitize=memory -fsanitize-memory-track-origins'\
  'RUSTFLAGS=-Zsanitizer=memory -Zsanitizer-memory-track-origins'\
  'RUSTDOCFLAGS=-Zsanitizer=memory -Zsanitizer-memory-track-origins'
cargo +nightly -Zbuild-std run --target x86_64-unknown-linux-gnu
```

Generally using MemorySanitizer can be quite challenging since all dependencies need to be instrumented.

Valgrind 3.16.0 improved situation with respect to false positives in memcheck due to compiler optimizations, so that might be an easier to use alternative.

**Comment by alex** [comment]
I think the conclusion here is that this is not a rustc bug, and this issue can be closed.
